func_adl/ast/function_simplifier.py

```diff
@@ -18,12 +22,56 @@ def arg_name():
     return n
 
 
+def make_args_unique(a: ast.Lambda) -> ast.Lambda:
+    '''
+    Replaces the lambda with a new lambda, with unique arguments names
+
+    Args:
+        a       Lambda function to be copied over
+
+    Returns:
+        l       New copy of the lambda. The original is unmodified.
+    '''
+    class replace_args (ast.NodeTransformer):
+        def __init__(self):
+            ast.NodeTransformer.__init__(self)
+            self._arg_stack: List[Tuple[str, str]] = []
+            self._seen_lambda = False
+
+        def visit_Lambda(self, node: ast.Lambda) -> ast.Lambda:
+            if self._seen_lambda:
+                mapping = [(a.arg, a.arg) for a in node.args.args]
+            else:
+                mapping = [(a.arg, arg_name()) for a in node.args.args]
+                self._seen_lambda = True
+
+            for old, new in mapping:
+                self._arg_stack.append((old, new))
+
+            r = self.generic_visit(node)
+            assert isinstance(r, ast.Lambda)
+
+            r.args.args = [ast.arg(arg=new, annotation=None) for old, new in mapping]
+            for arg in node.args.args:
+                self._arg_stack.pop()
+
+            return r
+
+        def visit_Name(self, node: ast.Name) -> ast.Name:
+            for n in reversed(self._arg_stack):
+                if n[0] == node.id:
+                    return ast.Name(id=n[1])
+            return node
+
+    return replace_args().visit(copy.deepcopy(a))
+
+
 def convolute(ast_g: ast.Lambda, ast_f: ast.Lambda):
     'Return an AST that represents g(f(args))'
     # TODO: fix up the ast.Calls to use lambda_call if possible
 
     # Combine the lambdas into a single call by calling g with f as an argument
-    l_g = copy.deepcopy(lambda_unwrap(ast_g))
+    l_g = make_args_unique(lambda_unwrap(ast_g))
     l_f = copy.deepcopy(lambda_unwrap(ast_f))
```

Comment:
Why is `make_args_unique()` run only on the outer lambda and not on the inner one?

Reply:
This is a good point - I'd not been creative enough in thinking about possible arguments, so I'd not implemented this. But there is no reason not to do it. Fixed in a commit to both of them use make_args_unique.